### PR TITLE
[USRMGR] Get text correctly even if the length is zero

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

JIRA issue: [CORE-18275](https://jira.reactos.org/browse/CORE-18275)

## Proposed changes

- Add `GetDlgItemTextAlloc` and `GetComboBoxLBTextAlloc` helper functions.
- Get the text data correctly even if the text length is zero.

## TODO

- [x] Do tests.

## Screenshots

BEFORE:
before

AFTER:
after

### CODE REVIEW

**Comment by HBelusca** [comment]
Is the difference in the "Before" screenshot about this "TEST" Full name also showing in the Administrator account?

File: dll/cpl/usrmgr/usrmgr.c [hunk 1]
```diff
@@ -15,6 +16,22 @@ static LONG APIENTRY UsrmgrApplet(HWND hwnd, UINT uMsg, LPARAM wParam, LPARAM lP
 
 HINSTANCE hApplet = 0;
 
+VOID GetDlgItemTextAlloc(HWND hwndDlg, INT nDlgItem, LPTSTR *ppsz)
+{
+    INT nLength = GetWindowTextLength(GetDlgItem(hwndDlg, nDlgItem));
+    *ppsz = HeapAlloc(GetProcessHeap(), 0, (nLength + 1) * sizeof(TCHAR));
+    if (*ppsz)
+        GetDlgItemText(hwndDlg, nDlgItem, *ppsz, nLength + 1);
+}
```
Comment (HBelusca) [inline dll/cpl/usrmgr/usrmgr.c:25]: For example:
```suggestion
LPTSTR GetDlgItemTextAlloc(HWND hwndDlg, INT nDlgItem)
{
    INT nLength = GetWindowTextLength(GetDlgItem(hwndDlg, nDlgItem));
    LPTSTR psz = HeapAlloc(GetProcessHeap(), 0, (nLength + 1) * sizeof(TCHAR));
    if (psz)
        GetDlgItemText(hwndDlg, nDlgItem, psz, nLength + 1);
    return psz;
}
```

**Comment by katahiromz** [comment]
It was unable to set an empty string.

File: dll/cpl/usrmgr/userprops.c [hunk 2]
```diff
@@ -87,80 +87,38 @@ SetUserProfileData(HWND hwndDlg,
...
-            nLength = SendMessage(GetDlgItem(hwndDlg, IDC_USER_PROFILE_DRIVE), CB_GETLBTEXTLEN, nIndex, 0);
-            pszHomeDrive = HeapAlloc(GetProcessHeap(), 0, (nLength + 1) * sizeof(TCHAR));
-            SendMessage(GetDlgItem(hwndDlg, IDC_USER_PROFILE_DRIVE), CB_GETLBTEXT, nIndex, (LPARAM)pszHomeDrive);
-            pUserInfo->usri3_home_dir_drive = pszHomeDrive;
+            GetComboBoxLBTextAlloc(hwndDlg, IDC_USER_PROFILE_DRIVE, nIndex,
+                                   &pUserInfo->usri3_home_dir_drive);
+        }
+        else
+        {
+            pUserInfo->usri3_home_dir_drive = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY,
```
Comment (HBelusca) [inline dll/cpl/usrmgr/userprops.c:120]: Isn't the existing code able to deal with NULL pointer strings so as to mean "empty string" ?

File: dll/cpl/usrmgr/usrmgr.c (hunk 1 above)
Comment (HBelusca) [reply dll/cpl/usrmgr/usrmgr.c:25]: Also, what to do if the window text length is zero? Should we always allocate a "one-character"-sized string just to store the NULL char? Or should we return NULL?

File: dll/cpl/usrmgr/usrmgr.c [hunk 3]
```diff
@@ -15,6 +16,22 @@ static LONG APIENTRY UsrmgrApplet(HWND hwnd, UINT uMsg, LPARAM wParam, LPARAM lP
 
 HINSTANCE hApplet = 0;
 
+VOID GetDlgItemTextAlloc(HWND hwndDlg, INT nDlgItem, LPTSTR *ppsz)
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/cpl/usrmgr/usrmgr.c]: Fixed.

File: dll/cpl/usrmgr/usrmgr.c (hunk 1 above)
Comment (katahiromz) [reply dll/cpl/usrmgr/usrmgr.c]: Fixed.

> Also, what to do if the window text length is zero?

This data is used in `dll/win32/netapi32/user.c`. `NULL` data is invalid. The empty string is valid data.

File: dll/cpl/usrmgr/userprops.c (hunk 2 above)
Comment (katahiromz) [reply dll/cpl/usrmgr/userprops.c:120]: `NULL` is an invalid data. The empty string is valid.

File: dll/cpl/usrmgr/groupprops.c [hunk 4]
```diff
@@ -494,32 +494,19 @@ SetGeneralGroupData(HWND hwndDlg,
...
 
     status = NetLocalGroupSetInfo(NULL, pGroupData->szGroupName, 1, (LPBYTE)&groupInfo, &dwIndex);
     if (status != NERR_Success)
     {
         ERR("NetLocalGroupSetInfo failed. Status: %lu  Index: %lu", status, dwIndex);
     }
 
-    if (pszComment)
-        HeapFree(GetProcessHeap(), 0, pszComment);
+    HeapFree(GetProcessHeap(), 0, &groupInfo.lgrpi1_comment);
```
Comment (HBelusca) [inline dll/cpl/usrmgr/groupprops.c:509]: bugfix
```suggestion
    HeapFree(GetProcessHeap(), 0, groupInfo.lgrpi1_comment);
```

File: dll/cpl/usrmgr/userprops.c [hunk 5]
```diff
@@ -87,80 +87,38 @@ SetUserProfileData(HWND hwndDlg,
...
-            pszHomeDrive = HeapAlloc(GetProcessHeap(), 0, (nLength + 1) * sizeof(TCHAR));
-            SendMessage(GetDlgItem(hwndDlg, IDC_USER_PROFILE_DRIVE), CB_GETLBTEXT, nIndex, (LPARAM)pszHomeDrive);
-            pUserInfo->usri3_home_dir_drive = pszHomeDrive;
+            pUserInfo->usri3_home_dir_drive =
+                GetComboBoxLBTextAlloc(hwndDlg, IDC_USER_PROFILE_DRIVE, nIndex);
+        }
+        else
+        {
+            pUserInfo->usri3_home_dir_drive = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY,
+                                                        sizeof(TCHAR));
```
Comment (HBelusca) [inline dll/cpl/usrmgr/userprops.c:121]: just place it in the same heapalloc line (and if you consider it too long then maybe put the whole HeapAlloc(....) call on a separate new line as you did for GetComboBoxLBTextAlloc above.

Comment (katahiromz) [reply dll/cpl/usrmgr/userprops.c]: Fixed.

**Review by JoachimHenze — APPROVED** [review]

**Pull request merged**